### ISSUE TITLE
A patterns test passes on the VM, but fails on dart2js

### ISSUE THREAD

**Issue by chloestefantsova** [dart-lang/sdk#50871]
The test in question is https://github.com/dart-lang/co19/blob/master/LanguageFeatures/Patterns/constant_A01_t04.dart

The printed error message is as follows:

```
Runtime error:
Expect.equals(expected: <+>, actual: <default>) fails.
```

It seems that the two symbol objects, one created as `Symbol("#")` in line 74 (https://github.com/dart-lang/co19/blob/ac4a905460672efd4006f633ea63d978ad577bf1/LanguageFeatures/Patterns/constant_A01_t04.dart#L74) and the other as a Symbol literal `#+` in line 34 (https://github.com/dart-lang/co19/blob/ac4a905460672efd4006f633ea63d978ad577bf1/LanguageFeatures/Patterns/constant_A01_t04.dart#L34), aren't treated as equal.

The test passes on the VM bot.

The command to run the dart2js test is as follows:

```
python3 tools/test.py -n dart2js-hostasserts-strong-linux-x64-chrome co19/LanguageFeatures/Patterns/constant_A01_t04
```

The command to run the VM test is as follows:

```
python3 tools/test.py -n dartk-strong-mac-release-x64 co19/LanguageFeatures/Patterns/constant_A01_t04
```

The failure needs more investigation to figure out if it's related to the CFE or to dart2js. I tentatively apply `area-front-end` label.

**Comment by sigmundch** [comment]
Coincidentally I was looking at this test yesterday. I am under the impression this is an issue with the test, not with the feature.

In particular, the test only fails on test1, not test2 or test3. The code in `test1` is using regular switch tables w/o records or patterns. 

I'm not sure how switch tables work in the VM, but either symbols are canonicalized or they are using some form of equality on the case tests. Switch tables in dart2js (and ddc for that matter) lower to switch tables in JavaScript, which do not use `==` to check for cases, but rely on keys being identical. Here the keys are not identical.

In other words, we see the VM/web discrepancy in this simple example w/o records or patterns:
```dart
String test(Symbol value) {
  switch (value) {
    case #foo:
      return "foo";
    default:
      return "default";
  }
}

main() {                              // On        dart2js  | VM
  print(test(Symbol("foo")));         // prints    default  | foo
  print(test(const Symbol("foo")));   // prints    foo      | foo
}
```

**Comment by johnniwinther** [comment]
This hits the underlying semantic difference between switch in <3.0 and >=3.0, the former (can) use `identical` (no user provided `operator ==`) and the latter use == (allow user provided `operator ==`).

**Comment by sigmundch** [comment]
Oh, that is a subtle change I wasn't aware of - that regular switch statements have new semantics.

Should we expect the semantic gap to always be addressed here by the CFE or do you foresee anything needed by each backend? For example, would we need to prevent the default lowering on certain cases to ensure they are as performant as they were before?

cc @rakudrama @alexmarkov @nshahan - in case there may be nuances for each of our backends that I may be missing.

**Comment by johnniwinther** [comment]
While the CL fixes the semantic issue, it is the plan to optimize the simple cases (like enum-switches) to use `SwitchStatement` so existing backend optimizations can be used.

**Comment by sigmundch** [comment]
Thanks @johnniwinther! I wonder whether there will be scenarios that may benefit from whole-program knowledge (e.g. an open interface used in a switch, whose type may or may not override `==` (e.g. only known to be true when computing the closed world)
